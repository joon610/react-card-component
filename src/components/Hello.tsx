import React from 'react'

export const Hello: React.FC<{text:string}> = ({text} ) =>{
  return (<h1>{text}</h1>)
}