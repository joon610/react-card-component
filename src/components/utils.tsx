export const objectToCssStyle = (object:object|undefined) => (JSON.stringify(object).replace(/"/g,'').replace(/((:{))/g,'{').replace(/,/g,';').slice(1,-1))
