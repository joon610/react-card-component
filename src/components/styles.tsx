import { css } from '@emotion/react'
import {objectToCssStyle} from './utils'

export const css_root = css`
  width: 100%;
  height:100%;
  background:white;
  color:dimgray;
`

export const css_child = css`
  height:100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
`

export const css_border = (hasFlg:boolean|undefined) => hasFlg && css`border-radius: 10px;`

export const css_outline = (hasFlg:boolean|undefined) => hasFlg && css`border: 1px solid #00000010;`

export const css_pointer = (hasFlg:boolean|undefined) => hasFlg && css`cursor:pointer;`

export const css_background = (color:string|undefined) => css`background:${color||`#f5f5f5`};`

export const css_user_style = (style:object|undefined) => css`${objectToCssStyle(style)}`

export const css_hover_option =(type:string|undefined)=> {
  switch (type) {
    case 'up':
      return css_hover_up;
    case 'down':
      return css_hover_down;
    case 'left':
      return css_hover_left;
    case 'right':
      return css_hover_right;
    case 'zoom':
      return css_hover_zoom;
    default:
      break;
  }
}

export const css_glass_option =  (hasFlg:boolean|undefined) => hasFlg && css_glass;

const css_glass = css`
	position: relative;
	z-index: 1;
	background: inherit;
	overflow: hidden;
  &:before {
    content: "";
    position: absolute;
    background: inherit;
    z-index: -1;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    box-shadow: inset 0 0 2000px rgba(255, 255, 255, .5);
    filter: blur(10px);
    margin: -20px;
  }

`


const css_transition = css`transition: background 150ms ease-out, transform 150ms ease-out;`

const css_hover_up = css`
  ${css_transition}
  :hover{
    border-color: #1EA7FD50;
    box-shadow: rgba(0, 0, 0, 0.06) 0 3px 10px 0;
    transform: translate3d(0, -4px, 0);
  }
`
const css_hover_down = css`
  ${css_transition}

  :hover{
    border-color: #1EA7FD50;
    box-shadow: rgba(0, 0, 0, 0.06) 0 3px 10px 0;
    transform: translate3d(0, 4px, 0);
  }
`
const css_hover_left = css`
  ${css_transition}
  :hover{
    border-color: #1EA7FD50;
    box-shadow: rgba(0, 0, 0, 0.06) 0 3px 10px 0;
    transform: translate3d(-4px, 0, 0);
  }
`
const css_hover_right = css`
  ${css_transition}
  :hover{
    border-color: #1EA7FD50;
    box-shadow: rgba(0, 0, 0, 0.06) 0 3px 10px 0;
    transform: translate3d(4px, 0, 0);
  }
`
const css_hover_zoom = css`
  ${css_transition}
  :hover{
    z-index:10;
    border-color: #1EA7FD50;
    box-shadow: rgba(0, 0, 0, 0.06) 0 3px 10px 0;
    transform: scale(1.1);
  }
`

