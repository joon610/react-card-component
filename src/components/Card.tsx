/** @jsx jsx */
import { css,jsx } from '@emotion/react'
import React from 'react'
import {css_root,css_child, css_user_style ,css_hover_option, css_pointer, css_border,css_outline, css_background} from './styles';
export interface BasicCardProps {
  children?: JSX.Element;
  className?: string;
  style?: object;
  bordered?: boolean;
  outlined?: boolean;
  background?: string;
  hoverType?: 'up' | 'left' | `right` | `down` | 'zoom' | undefined ;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
}

export const Card:React.FC<BasicCardProps> = ({children,background, outlined=false, bordered=false, style={}, hoverType, ...props}) => {

  const css_combine = [
    css_root,
    css_outline(outlined),
    css_border(bordered),
    css_background(background),
    css_pointer(!!props?.onClick),
    css_hover_option(hoverType),
    css_user_style(style)
  ];
  return (
  <div css={css_combine} {...props}>
    <div css={css_child}>
      {children}
    </div>
  </div>
  )
}