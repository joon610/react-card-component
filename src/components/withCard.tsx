/** @jsx jsx */
import { css,jsx } from '@emotion/react'
import React from 'react'
import {css_elevation, css_hover_option} from './styles';
interface BasicCardProps {
  children: JSX.Element
  className?: string
  elevation?: elevationType
  radius?: string;
  hoverOption?: hoverMoveType | undefined;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void | undefined;
}

type elevationType =  0 | 1 | 2
type hoverMoveType = 'up' | 'left' | `right` | `down`


const css_root = css`
  width: 100%;
  height:100%;
  border-radius: 5px;
`

const css_child = css`
  height:100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
`


const BasicCard = (Component:React.FC<{ children?:JSX.Element}>) => (props:BasicCardProps) => {
  const elevationValue:elevationType =  props?.elevation || 0;
  const css_hover = props?.hoverOption && css_hover_option[props?.hoverOption];
  return (
  <div css={[css_root,css_elevation[elevationValue],css_hover]} className={props?.className} onClick={props.onClick}>
    <div css={css_child}>
      <Component { ...props}/>
    </div>
  </div>
  )
}

export default BasicCard;