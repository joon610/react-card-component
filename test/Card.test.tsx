import React from 'react'
import { fireEvent, render } from "@testing-library/react";
import { matchers } from '@emotion/jest'
import '@testing-library/jest-dom'
import {Card} from '../src/components/Card'
import { objectToCssStyle } from '../src/components/utils';
import { relative } from 'path';
import { inherits } from 'util';
expect.extend(matchers)

describe('<Card>',()=>{
  it('empty snapshot', () => {
    const {container} = render(<Card/>);
    expect(container).toMatchSnapshot();
  });

  it('Rendering Test', ()=>{
    const {getByText} = render(<Card><div>good</div></Card>);
    const childText = getByText('good');
    expect(childText).toBeInTheDocument();
  })

  it('Background', ()=>{
    const {container} = render(<Card background={"red"}><div>good</div></Card>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle('background:red');
    expect(container).toMatchSnapshot();
  })

  it('css Object style',()=>{
    const { container } =render(<Card style={{"width":"300px","background":"red",":hover":{"background":"blue"}}}/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle('width:300px');
    expect(firstChild).toHaveStyle('background:red');
    expect(firstChild).toHaveStyleRule("background", "blue", { target: ":hover" });
    expect(container).toMatchSnapshot();
  })

  it('click Card', ()=>{
    const handleClick = jest.fn();
    const { container } =render(<Card  onClick={handleClick}/>);
    const firstChild = container.getElementsByTagName('div')[0];
    fireEvent.click(firstChild);
    expect(firstChild).toHaveStyle('cursor:pointer')
    expect(handleClick).toHaveBeenCalledTimes(1);
  })

  it('bordered ',()=>{
    const { container } =render(<Card bordered/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle({'border-radius':'10px'})
    expect(container).toMatchSnapshot();
  })

  it('outlined ',()=>{
    const { container } =render(<Card outlined/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle({'border':'1px solid #00000010'})
    expect(container).toMatchSnapshot();
  })

  it('outlined && bordered',()=>{
    const { container } =render(<Card outlined bordered/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle({'border':'1px solid #00000010','border-radius':'10px'});
    expect(container).toMatchSnapshot();
  })

  it('hover zoom', async ()=>{
    const { container } =render(<Card hoverType={"zoom"}/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("z-index", "10", { target: ":hover" });
    expect(firstChild).toHaveStyleRule("transform", "scale(1.1)", { target: ":hover" });
  })

  it('hover up', async ()=>{
    const { container } =render(<Card hoverType={"up"}/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate3d(0, -4px, 0)", { target: ":hover" });
  })

  it('hover down', async ()=>{
    const { container } =render(<Card hoverType={"down"}/>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate3d(0, 4px, 0)", { target: ":hover" });
  })

  it('hover left', async ()=>{
    const { container } =render(<Card hoverType={"left"}><div>good</div></Card>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate3d(-4px, 0, 0)", { target: ":hover" });
  })

  it('hover right', async ()=>{
    const { container } =render(<Card hoverType={"right"}><div>good</div></Card>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate3d(4px, 0, 0)", { target: ":hover" });
  })

  it('Glass', async ()=>{
    const { container } =render(<Card glass><div>good</div></Card>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle({'position':'relative', 'z-index':'1',  'overflow':'hidden'});
    expect(firstChild).toHaveStyleRule("content", '""', { target: ":before" });
    expect(firstChild).toHaveStyleRule("position", "absolute", { target: ":before" });
    expect(firstChild).toHaveStyleRule("z-index", "-1", { target: ":before" });
    expect(firstChild).toHaveStyleRule("top", "0", { target: ":before" });
    expect(firstChild).toHaveStyleRule("left", "0", { target: ":before" });
    expect(firstChild).toHaveStyleRule("right", "0", { target: ":before" });
    expect(firstChild).toHaveStyleRule("bottom", "0", { target: ":before" });
    expect(firstChild).toHaveStyleRule("box-shadow", "inset 0 0 2000px rgba(255, 255, 255, .5)", { target: ":before" });
    expect(firstChild).toHaveStyleRule("margin", "-20px", { target: ":before" });
    expect(firstChild).toHaveStyleRule("filter", "blur(10px)", { target: ":before" });
  })
})

describe('utils',()=>{
  it('objectToCssStyle',()=>{
    const styleObject = {
      width:"300px",
      height:"200px",
      ":hover":{
        background:"red"
      }
    }

    expect((objectToCssStyle(styleObject))).toEqual('width:300px;height:200px;:hover{background:red}')
  })
})