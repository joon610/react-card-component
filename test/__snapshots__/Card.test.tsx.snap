// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card> Background 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  background: red;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        good
      </div>
    </div>
  </div>
</div>
`;

exports[`<Card> bordered  1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  border-radius: 10px;
  background: #f5f5f5;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Card> css Object style 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  background: #f5f5f5;
  width: 300px;
  background: red;
}

.emotion-0:hover {
  background: blue;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Card> empty snapshot 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  background: #f5f5f5;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Card> outlined  1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  border: 1px solid #00000010;
  background: #f5f5f5;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Card> outlined && bordered 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  background: white;
  color: dimgray;
  border: 1px solid #00000010;
  border-radius: 10px;
  background: #f5f5f5;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
  </div>
</div>
`;
